ledger/src/shred.rs

```diff
@@ -811,129 +811,54 @@ impl Shredder {
-                return Err(reed_solomon_erasure::Error::TooFewShardsPresent);
+        if num_coding == 0 || shreds.len() >= fec_set_size {
+            return Ok(Vec::default());
+        }
+        let mut mask = vec![false; num_data];
+        let mut blocks = vec![None; fec_set_size];
+        for mut shred in shreds {
+            if (shred.index() as usize) < first_index {
+                return Err(InvalidIndex);
             }
-
-            let session = Session::new(num_data, num_coding)?;
-
-            // All information (excluding the restricted section) from a data shred is encoded
-            let valid_data_len = SHRED_PAYLOAD_SIZE - SIZE_OF_CODING_SHRED_HEADERS;
-            let coding_block_offset = SIZE_OF_CODING_SHRED_HEADERS;
-            let mut blocks: Vec<(&mut [u8], bool)> = shred_bufs
-                .iter_mut()
-                .enumerate()
-                .map(|(position, x)| {
-                    if position < num_data {
-                        x[..valid_data_len].as_mut()
-                    } else {
-                        x[coding_block_offset..].as_mut()
-                    }
-                })
-                .zip(present.clone())
-                .collect();
-            session.decode_blocks(&mut blocks)?;
-
-            let mut num_drained = 0;
-            present
-                .iter()
-                .enumerate()
-                .for_each(|(position, was_present)| {
-                    if !*was_present && position < num_data {
-                        let drain_this = position - num_drained;
-                        let shred_buf = shred_bufs.remove(drain_this);
-                        num_drained += 1;
-                        if let Ok(shred) = Shred::new_from_serialized_shred(shred_buf) {
-                            let shred_index = shred.index() as usize;
-                            // Valid shred must be in the same slot as the original shreds
-                            if shred.slot() == slot {
-                                // A valid data shred must be indexed between first_index and first+num_data index
-                                if (first_index..first_index + num_data).contains(&shred_index) {
-                                    recovered_data.push(shred)
-                                }
-                            }
-                        }
-                    }
-                });
+            let offset = if shred.is_data() { 0 } else { num_data };
+            let index = offset + shred.index() as usize - first_index;
+            if shred.is_data() {
+                if index >= num_data {
+                    return Err(InvalidIndex);
+                }
+                mask[index] = true;
+                shred.payload.resize(CODE_BLOCK_SIZE, 0u8);
+            } else {
+                if index >= fec_set_size {
+                    return Err(InvalidIndex);
+                }
+                shred.payload.drain(..SIZE_OF_CODING_SHRED_HEADERS);
```

Comment:
Maybe a comment here: 

"Ignore the headers of coding shreds as only the body of the shred is needed for symbol recovery"

Reply:
done